### PR TITLE
Various C/C++ fixes

### PR DESCRIPTION
**Pull request by Manishearth**
- Correctly handle `&str` fields in structs
 - correctly handle returning structs in results
 - correctly handle non-opaque receivers in the C++ backend

I'm not updating tests for this because @QnnOkabayashi has some pending JS work that would also involve such tests.

This PR can be thoroughly tested if:

 - a struct that is being taken as an input has a str field
 - a struct that is being returned in a result has a str field
 - an input struct has a str field
 - a struct has a `fn consume(self)` method on it
 
Fixes https://github.com/rust-diplomat/diplomat/issues/175

### CODE REVIEW

File: tool/src/cpp/types.rs [hunk 1]
```diff
@@ -119,7 +127,12 @@ fn gen_type_inner<W: fmt::Write>(
         }
 
         ast::TypeName::StrReference(_) => {
-            write!(out, "const {}", library_config.string_view.expr)?;
+            let maybe_const = if in_struct {
+                ""
+            } else {
+                "const "
+            };
+            write!(out, "{maybe_const}{}", library_config.string_view.expr)?;
```
Comment (QnnOkabayashi) [inline tool/src/cpp/types.rs]: nit: just put a `write!` in each if branch?

**Review by QnnOkabayashi — APPROVED** [review]
lgtm! Unrelated but it looks like the C++ backend could use some love :(

File: tool/src/cpp/types.rs (hunk 1 above)
Comment (Manishearth) [reply tool/src/cpp/types.rs]: I personally find it cleaner to have fewer writes

File: feature_tests/js/test/struct.mjs [hunk 2]
```diff
@@ -3,11 +3,12 @@ import test from 'ava';
...
 test("Verify invariants of struct", t => {
-    const s = MyStruct.new();
+    const s = MyStruct.new("hello");
     t.is(s["a"], 17);
     t.is(s["b"], true);
     t.is(s["c"], 209);
     t.is(s["d"], 1234n);
     t.is(s["e"], 5991);
     t.is(s["f"], "餐");
+    // t.is(s["g"], "hello");
```
Comment (Manishearth) [inline feature_tests/js/test/struct.mjs:13]: @QnnOkabayashi this test doesn't currently work, but I added it anyway for future use

**Pull request merged**